keps/2724-topology-aware-schedling/README.md

```diff
@@ -0,0 +1,672 @@
+  // - Require set when `kueue.x-k8s.io/require-topology` annotation is set on the Job
+  Policy TopologyRequestPolicy
+
+  // Level indicated by the `kueue.x-k8s.io/prefer-topology` or `kueue.x-k8s.io/require-topology `
+  // annotation
+  Level string
+}
+```
+
+We extend the `PodSetAssignment` structure to keep track of the number of pods
+at each topology level to the specific subset of nodes.
+
+```golang
+type PodSetAssignment struct {
+  ...
+
+  // TopologyAssignment indicates the resources assigned per topology level
+  TopologyAssignment *TopologyAssignment
+}
+
+type TopologyAssignment struct {
+     // Slices contains the list of assignments split into slices
+     Slices []TopologyAssignmentSlice
+}
+
+type TopologyAssignmentSlice struct {
+  // NodeLabels constitutes the nodeSelector for a given slice of pods. It
+  // defines values for all labels configured in the Topology.Levels.
+  NodeLabels map[string]string
+
+  // Count indicates the number of pods in a given TopologyAssignmentSlice
+  Count int
+}
+```
+
+Kueue uses the `kueue.x-k8s.io/topology` scheduling gate to delay the
+`nodeSelector` assignment, because different pods in the same PodSet may have
+different values:
+
+```golang
+const (
+  // TopologySchedulingGate is used to delay topology assignment for pods
+  // once all the pods are created.
+  TopologySchedulingGate = "kueue.x-k8s.io/topology"
+
+  // WorkloadAnnotation indicates the name of the workload assigned.
+  WorkloadAnnotation = "kueue.x-k8s.io/workload"
+
+  // PodSetLabel indicates the name of the PodSet in the workload
+  PodSeLabel = "kueue.x-k8s.io/podset"
+)
+```
+
+### Computing the assignment
+
+The extended pod assignment is set on admitting the workload. In order to
+compute the assignment Kueue caches the information about the node allocatable
+capacity and the current usage by other workloads in this resource flavor.
+
+The information about the allocatable capacity is scraped from nodes based on
+the `status.allocatable` field.
+
+The cached information about the used resources is updated whenever a workload
+is admitted, suspended, resumed or finished. Additionally, we scrape the usage
+information from all non-TAS pods bound to the TAS nodes - this is done to
+account for Pods not created by Workloads not managed by Kueue (such as
+DaemonSet pods).
+
+The available capacity for TAS nodes is then computed as a difference between
+the allocatable space and current usage.
+
+For a given PodSet Kueue:
+- when the `require-topology` is used, then Kueue tries to find any value of the
+ level label which can accommodate all the pods. If there is no such value, then
+ the workload keeps waiting in the queue.
+- when the `prefer-topology` is used, then Kueue tries to use heuristics to
+  minimize the number of values of the label used. One approach is to greedily
+  choose the value which can accommodate the most pods.
+
+### Enforcing the assignment
```

Comment:
What happens when the TAS and batch user-specified scheduling directives like nodSelector, topologySpreadConstrains, and affinity are semantically conflicted?

I guess that Pod Infinity is stuck, right? Or would you like to prevent specifying the scheduling directives by batch users?

Reply:
TAS only uses regular "nodeSelector" so no difference here compared to "regular" ResourceFlavors. The only difference is that TAS might inject a different nodeSelector for a subset of pods in a Job.

As in regular Kueue, if you use some pod affinity conflicting with the injected nodeSelector (or you miss some tolerations), then the pods get stuck, a user can use waitForPodsReady to mitigate it.